src/server/array_func.c

```diff
@@ -653,32 +652,17 @@ setup_arrayjob_attrs(attribute *pattr, void *pobj, int mode)
 	pjob->ji_wattr[(int)JOB_ATR_array].at_flags = ATR_VFLAG_SET | ATR_VFLAG_MODCACHE;
 	job_attr_def[(int)JOB_ATR_array_state_count].at_free(&pjob->ji_wattr[(int)JOB_ATR_array_state_count]);
 
-	if (mode == ATR_ACTION_RECOV) {
-		int x, y, z, ct;
-		char *ep;
-
-		/* on recovery ... */
-		/* parse the various components again, since we dont store them */
-		if (parse_subjob_index(pattr->at_val.at_str, &ep, &x, &y, &z, &ct) != 0)
+	if ((mode == ATR_ACTION_NEW) || (mode == ATR_ACTION_RECOV)) {
+		if (pjob->ji_ajtrk)
+			free(pjob->ji_ajtrk);
+		if ((pjob->ji_ajtrk = mk_subjob_index_tbl(pjob->ji_wattr[(int)JOB_ATR_array_indices_submitted].at_val.at_str, JOB_STATE_QUEUED)) == NULL)
 			return PBSE_BADATVAL;
+	}
 
-		pjob->ji_ajtrk->tkm_ct = ct;
-		pjob->ji_ajtrk->tkm_step = z;
-		pjob->ji_ajtrk->tkm_flags = 0;
-
-		/* reset counts and any running/exiting subjob to queued */
-		for (i=0; i < PBS_NUMJOBSTATE; ++i)
-			pjob->ji_ajtrk->tkm_subjsct[i] = 0;
-		for (i=0; i < pjob->ji_ajtrk->tkm_ct; ++i) {
-			if ((pjob->ji_ajtrk->tkm_tbl[i].trk_status == JOB_STATE_RUNNING)
-				|| (pjob->ji_ajtrk->tkm_tbl[i].trk_status == JOB_STATE_EXITING))
-				pjob->ji_ajtrk->tkm_tbl[i].trk_status =JOB_STATE_QUEUED;
-			pjob->ji_ajtrk->tkm_subjsct[pjob->ji_ajtrk->tkm_tbl[i].trk_status]++;
-		}
-
-		/* clear and reset array_indices_remaining to new value */
-		job_attr_def[(int)JOB_ATR_array_indices_remaining].at_free(&pjob->ji_wattr[(int)JOB_ATR_array_indices_remaining]);
-		job_attr_def[(int)JOB_ATR_array_indices_remaining].at_decode(&pjob->ji_wattr[(int)JOB_ATR_array_indices_remaining], NULL, NULL, cvt_range(pjob->ji_ajtrk, JOB_STATE_QUEUED));
+	if (mode == ATR_ACTION_RECOV) {
```

Comment:
I think this if block has to be moved to the line number 688 where the value of JOB_ATR_array_indices_remaining is actually set. Also for setting the attribute value using the right method is to use the function set_attr_svr which is recently introduced.

Reply:
JOB_ATR_array_indices_remaining kind of stands out wrt other JOB_ATR_xxx atribs operations. I think this is done to speed up things for job arrays. JOB_ATR_array_indices_remaining is only set in status_job() here https://github.com/PBSPro/pbspro/blob/master/src/server/stat_job.c#L266-L272 and is marked for updation using the ATR_VFLAG_MODCACHE at various places. We wont be decoding for ATR_ACTION_RECOV and returning. Hence this if block code is at proper place and  cannot be moved to 688 as explained.